src/python-bindings/htcondor2/collector.cpp

```diff
@@ -1 +1,289 @@
+			Py_DecRef(pyClassAd);
+
+			if(rv != 0) {
+				// PyList_Append() has already set an exception for us.
+				return NULL;
+			}
+		}
+
+		return list;
+	} else {
+		switch (result) {
+			case Q_COMMUNICATION_ERROR:
+				// This was HTCondorIOError in version 1.
+				PyErr_SetString( PyExc_RuntimeError, "Failed communication with collector." );
+				return NULL;
+
+			case Q_INVALID_QUERY:
+				// This was HTCondorIOError in version 1.
+				PyErr_SetString( PyExc_RuntimeError, "Invalid query." );
+				return NULL;
+
+			case Q_NO_COLLECTOR_HOST:
+				// This was HTCondorLocateError in version 1.
+				PyErr_SetString( PyExc_RuntimeError, "Unable to determine collector host." );
+				return NULL;
+
+			// In version 1, we believe these errors were impossible.
+			case Q_PARSE_ERROR:
+			case Q_MEMORY_ERROR:
+			case Q_INVALID_CATEGORY:
+
+			default:
+				// This was HTCondorInternalError in version 1.
+				PyErr_SetString( PyExc_RuntimeError, "Unknown error from collector query." );
+				return NULL;
+		}
+	}
+}
+
+
+static PyObject *
+_collector_locate_local( PyObject *, PyObject * args ) {
+	// _collector_locate_local(self, _handle, daemon_type)
+
+	PyObject * self = NULL;
+	PyObject_Handle * handle = NULL;
+	daemon_t daemon_type = DT_NONE;
+	if(! PyArg_ParseTuple( args, "OOl", & self, (PyObject **)& handle, & daemon_type )) {
+		// PyArg_ParseTuple() has already set an exception for us.
+		return NULL;
+	}
+
+	Daemon local( daemon_type, 0, 0 );
+	if( local.locate() ) {
+		ClassAd * locateOnlyAd = local.locationAd();
+		if( locateOnlyAd == NULL ) {
+			PyErr_SetString( PyExc_RuntimeError, "Found local daemon but failed to acquire its ad." );
+			return NULL;
+		}
+
+		PyObject * pyClassAd = py_new_classad2_classad(locateOnlyAd->Copy());
+		// PyObject * list = Py_BuildValue( "[N]", pyClassAd );
+		return pyClassAd;
+	} else {
+		// This was HTCondorLocateError in version 1.
+		PyErr_SetString( PyExc_RuntimeError, "Unable to locate local daemon." );
+		return NULL;
+	}
+}
+
+
+static PyObject *
+_collector_advertise( PyObject *, PyObject * args ) {
+	// _collector_advertise( self._handle, ad_list, command, use_tcp )
+
+	PyObject_Handle * handle = NULL;
+	PyObject * py_list_of_ads = NULL;
+	const char * command_str = NULL;
+	int use_tcp = 1;
+	if(! PyArg_ParseTuple( args, "OOzp", (PyObject **)& handle, & py_list_of_ads, & command_str, & use_tcp)) {
+		// PyArg_ParseTuple() has already set an exception for us.
+		return NULL;
+	}
+
+	auto * collectorList = (CollectorList *)handle->t;
+
+	Py_ssize_t length = PyList_Size(py_list_of_ads);
+	if( length == 0 ) {
+		Py_RETURN_NONE;
+	}
+
+	int command = getCollectorCommandNum(command_str);
+	if( command == -1 ) {
+		// This was HTCondorEnumError in version 1.
+		PyErr_SetString( PyExc_ValueError, "invalid command" );
+		return NULL;
+	}
+	if( command == UPDATE_STARTD_AD_WITH_ACK ) {
+		PyErr_SetString( PyExc_NotImplementedError, "startd-with-ack protocol" );
```

Comment:
Should the message include the lack of implementation like V1 does?

Reply:
It hardly seems necessary, given that it's a `NotImplementedError`, but I'll change if it you think that's a good idea.

Reply:
Totally missed the fact that it was throwing a `NotImplementedErro` so I think that is alright